pkg/controller/webhook/webhook_certificates_controller.go

```diff
@@ -0,0 +1,114 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package webhook
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/certificates"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/watches"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	"k8s.io/api/admissionregistration/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ControllerName = "webhook-certificates-controller"
+)
+
+var log = logf.Log.WithName(ControllerName)
+
+var _ reconcile.Reconciler = &ReconcileWebhookResources{}
+
+// ReconcileWebhookResources reconciles the certificates used by the webhook server.
+type ReconcileWebhookResources struct {
+	// k8s.Client is used to watch resources
+	k8s.Client
+	// iteration is the number of times this controller has run its Reconcile method
+	iteration uint64
+
+	// webhook parameters
+	webhookParams Params
+	// resources are updated with a native Kubernetes client
+	clientset kubernetes.Interface
+}
+
+func (r *ReconcileWebhookResources) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	defer common.LogReconciliationRun(log, request, &r.iteration)()
+	if err := r.webhookParams.ReconcileResources(r.clientset); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Get the latest content of the webhook CA
+	webhookServerSecret, err := r.clientset.CoreV1().Secrets(r.webhookParams.Namespace).Get(r.webhookParams.SecretName, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	serverCA := certificates.BuildCAFromSecret(*webhookServerSecret)
+	if serverCA == nil {
+		return reconcile.Result{}, fmt.Errorf("cannot find CA in webhook secret %s/%s", r.webhookParams.Namespace, r.webhookParams.SecretName)
+	}
+
+	return reconcile.Result{
+		RequeueAfter: certificates.ShouldRotateIn(time.Now(), serverCA.Cert.NotAfter, r.webhookParams.Rotation.RotateBefore),
```

Comment:
We have a protection against timer leaks in https://github.com/elastic/cloud-on-k8s/blob/6e6118b3e1f40b2527912c7c8aa319f0943df1fd/pkg/controller/common/reconciler/results.go#L72-L77

As long as https://github.com/kubernetes/kubernetes/pull/83945 is not merged and released we should probably do something similar here.